### PR TITLE
[BugFix] Fix ModelOpt quantization inference for fused siblings

### LINKED ISSUE #47241: [Bug]: Nemotron-3-Super-120B NVFP4 generation degenerates on recent `main` — bisected to #46756
### Summary

On recent `main`, `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4` produces correct short answers but **degenerates into repetition on any multi-step / longer generation**. The smaller `nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-NVFP4` is **unaffected**. A `git bisect` points at **#46756 ("Add MiniMax-M3 modelopt nvfp4 support", `debec6440`)**, which modifies the shared ModelOpt-NVFP4 quantization path that these checkpoints route through.

We were validating NVFP4 on consumer/SoC Blackwell and ran into this — sharing the full investigation in case it's useful. Happy to run more tests; we have a reproducible setup.

### Your current environment

- vLLM `main` (~`0.22.1rc1.dev`, reproduced on commits `ac521f623` and `8cf7c4d8a`)
- torch `2.11.0+cu130`, flashinfer `0.6.12`/`0.6.13`, CUDA 13.0
- Reproduced on **two** GPUs, so it is **not hardware-specific**:
  - RTX PRO 6000 (sm120, discrete Blackwell)
  - GB10 / DGX Spark (sm121)
- Models: `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4` (affected), `nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-NVFP4` (not affected)
- `moe_backend=marlin`, `linear_backend=cutlass`, default fp8 KV — but see note below: the trigger is upstream of the MoE backend, so this is not backend-specific.

### 🐛 Describe the bug

Greedy decoding (`temperature=0`) on **Super** on recent `main`:

| prompt | output on recent `main` |
|---|---|
| `What is 2+2? Reply with just the number.` | `The answer is 4. The the answer is the.` |
| `Name the largest planet in our solar system.` | `...Jupiter... The largest planet in the the world is a giant planet...` |
| `A shop has 5 apples and sells 2. How many remain?` | `The answer is 2. The number of oranges is 1... 1000000000000...` |

Short factual answers are roughly right, but outputs collapse into repetition/nonsense as soon as a couple of reasoning steps are needed — so anything like GSM8K goes from ~1.0 to ~0.0. The **Nano** checkpoint answers all of these correctly on the same build.

We reproduced this identically on sm120 and sm121, which rules out a GB10-specific kernel.

### Bisect (showing our work)

We bisected the range from the last-known-good era to current `main` (397 commits) using a small, scriptable predicate: 8 grade-school word problems, scored `GOOD` if ≥6/8 correct and `BAD` if ≤3/8 (the signal is clean — good commits score 8/8, regressed commits score 1/8). Each step: `git checkout <commit>` → rebuild → load Super with `moe_backend=marlin` → score.

Confirmed boundary:

| commit | PR | subject | verdict |
|---|---|---|---|
| `07d33e575` | #44657 | [MyPy] Fix mypy incompatible assignment errors | **GOOD (8/8)** |
| `379acd4e4` | #46860 | [Bugfix][Quantization] Fix W8A8 int-quantized scheme selection regression | (int8 path — unrelated to NVFP4) |
| `c8fb2963b` | #46713 | [FS-Offloading] Batch Lookup in C | (filesystem offload — unrelated to numerics) |
| **`debec6440`** | **#46756** | **Add MiniMax-M3 modelopt nvfp4 support** | **BAD (1–2/8)** |
| `4708292d4` | #46683 | Bump flashinfer version to 0.6.13 | BAD |

Two things worth noting:
- **flashinfer was held constant (0.6.12) across the GOOD→BAD boundary** — the 0.6.13 bump is the *next* commit (#46683) — so this is a vLLM code change, not a dependency change.
- Of the three commits in the `07d33e575 → debec6440` window, **#46756 is the only one that touches the NVFP4 path**; the other two are an int8-quant fix and a filesystem-offloading change and can't plausibly affect NVFP4 generation.

### Suspected mechanism

#46756 changes:

```
 vllm/model_executor/layers/quantization/modelopt.py                | 27 +++++++++
 vllm/model_executor/layers/fused_moe/experts/trtllm_nvfp4_moe.py   | 68 ++++++++++++++++---
 vllm/model_executor/layers/quantization/utils/flashinfer_utils.py  |  1 +
 tests/quantization/test_modelopt.py                                |  6 ++
```

`modelopt.py` is the ModelOpt-NVFP4 quantization handler that **all** ModelOpt-NVFP4 checkpoints go through, and it sits *upstream of the MoE-backend selection* — which is why the regression shows up even on the `marlin` path, not just the trtllm one. Nemotron-3's checkpoints are ModelOpt-NVFP4.

Our best guess at why **Super** breaks while **Nano** doesn't is Super's extra structure — it uses a Latent-MoE (`moe_latent_size=1024`) and larger hidden size (4096 vs 2688) — so a change to how modelopt config / scale handling is resolved may interact with Super's layout in a way Nano's doesn't exercise. We haven't confirmed the exact line, and would defer to the #46756 authors on that.

### Reproduction

Minimal manual repro (any Blackwell GPU with ≥90 GB VRAM for Super):

```bash
# current main -> degenerates
pip install -e .  # at debec6440 or later
python -c "
from vllm import LLM, SamplingParams
llm = LLM('nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-NVFP4', moe_backend='marlin', enforce_eager=True, max_model_len=2048)
tok = llm.get_tokenizer()
p = tok.apply_chat_template([{'role':'user','content':'A shop has 5 apples and sells 2. How many remain? Answer with just the number.'}], tokenize=False, add_generation_prompt=True)
print(llm.generate([p], SamplingParams(temperature=0, max_tokens=48))[0].outputs[0].text)
"
# -> degenerates

# checkout debec6440^ (the parent of #46756) and repeat -> correct answer
```

We also have a reproducible container (current `main` prebuilt for sm120) and can share exact logs / rerun anything that would help narrow it further.

### Before submitting a new issue...

- [x] Searched existing issues.
- [x] Included a minimal reproduction and the bisected commit.

### PR DESCRIPTION
**Pull request by jasonlizhengjian**
## Summary

Fix ModelOpt mixed-precision dispatch so unlisted fused projections remain unquantized.

This removes the generic parent-sibling fallback from `ModelOptMixedPrecisionConfig._resolve_quant_algo` and adds explicit `packed_modules_mapping` declarations to the NVIDIA MiniMax-M3 causal and conditional model classes. Explicit packed mappings preserve MiniMax `qkv_proj` / `gate_up_proj` dispatch without inferring quantization for unrelated or intentionally unquantized siblings.

Fixes #47241.
Follow-up to #46756.

## Root cause

Qwen3.6 stores GDN `in_proj_a` and `in_proj_b` as BF16 tensors without quantization scales, while vLLM fuses them into `in_proj_ba`.

The parent-prefix fallback added by #46756 observed that the other direct `linear_attn` children were FP8 and incorrectly classified the unlisted `in_proj_ba` as FP8. The loader then copied BF16 checkpoint tensors into FP8 parameters while the expected scale placeholders had no checkpoint values. Shapes still matched, so loading completed, but recurrent attention state and logits were corrupted.

## Why this is not duplicate work

#47354 guards the fallback with `packed_modules_mapping`. That fixes unregistered siblings such as Nemotron Mamba `conv1d` and `norm`, but Qwen's unquantized `in_proj_ba` is itself a registered packed module, so the guarded fallback still classifies it as FP8.

This PR removes the unsafe inference entirely and makes the original MiniMax requirement explicit in the NVIDIA model classes. Resolution therefore succeeds only when declared packed shards are present and agree on their quantization algorithm.

## Validation

- B300 SXM6 AC, Qwen3.6-35B-A3B-NVFP4, deterministic prompts with `temperature=0` and `enable_thinking=false`:
  - Before: both prompts emitted token `248050` (`<|box_end|>`) repeatedly.
  - After: France prompt returned `Paris`; `2+2` returned `4`.
- B300 focused ModelOpt mixed-precision tests: `6 passed`.
- After rebasing onto current `main`, CPU-safe resolver regression tests: `4 passed`.
- `pre-commit run --files vllm/model_executor/layers/quantization/modelopt.py vllm/models/minimax_m3/nvidia/model.py tests/quantization/test_modelopt.py`: all applicable hooks passed.

## AI assistance

OpenAI Codex assisted with root-cause analysis, implementation, and test execution. The human submitter reviewed the complete diff and validation results.

### CODE REVIEW

**Comment by eugr** [comment]
Any progress on this? This PR fixes the underlying issue that breaks NVFP4 quants in current nightly.
@mgoin - fyi.

**Comment by javasparrows** [comment]
Confirming this issue independently on DGX Spark (GB10 / sm_121, aarch64, driver 580.159.03) with `nvidia/Qwen3.6-27B-NVFP4` — and that this PR's analysis matches what we observed.

**Symptom**: every request returns repeated `!` (token-0 collapse), on both `/v1/completions` and `/v1/chat/completions`, greedy or not. Decode throughput also drops (~9 tok/s vs ~23 expected).

**Image-level bisect** (docker `vllm/vllm-openai` nightlies):
- `nightly-a65f93fb...` (2026-06-28, pre-#46756): ✅ correct output
- `nightly-34b560b7...` (2026-07-05): ❌ repeated `!`
- `nightly-69715823...` (2026-07-06): ❌ repeated `!`

Consistent with the regression landing in #46756 (merged 06-29).

**Ruled out experimentally** (all still corrupted): MTP speculative config on/off, `--kv-cache-dtype fp8` vs `auto`, `VLLM_USE_FLASHINFER_SAMPLER` on/off, extra parser flags, checkpoint corruption (all shards sha256-verified against HF LFS). A BF16 model (Qwen3-0.6B) on the same broken images works fine, and a pure-NVFP4 (non-MIXED_PRECISION) community checkpoint of the same model also works — pointing squarely at the mixed-precision dispatch path this PR fixes.

Currently pinned to the 06-28 nightly as a workaround. Would love to see this merged — happy to test a wheel/image with this patch on sm_121 if useful.

**Review by vadiklyutiy — APPROVED** [review]

**Comment by jasonlizhengjian** [comment]
@mgoin should be good to merge
